include/pybind11/detail/internals.h

```diff
@@ -310,25 +310,38 @@ struct type_info {
 #    endif
 #endif
 
-/// On Linux/OSX, changes in __GXX_ABI_VERSION__ indicate ABI incompatibility.
-/// On MSVC, changes in _MSC_VER may indicate ABI incompatibility (#2898).
-#ifndef PYBIND11_BUILD_ABI
-#    if defined(__GXX_ABI_VERSION)
-#        define PYBIND11_BUILD_ABI "_cxxabi" PYBIND11_TOSTRING(__GXX_ABI_VERSION)
-#    elif defined(_MSC_VER)
-#        define PYBIND11_BUILD_ABI "_mscver" PYBIND11_TOSTRING(_MSC_VER)
+// Catch other conditions that imply ABI incompatibility
+// - MSVC builds with different CRT versions
+// - An anticipated MSVC ABI break ("vNext")
+// - Builds using libc++ with unstable ABIs
+// - Builds using libstdc++ with the legacy (pre-C++11) ABI
+#if defined(_MSC_VER)
+#    if defined(_MT) && defined(_DLL) // catches /MD or /MDd
+#        define PYBIND11_BUILD_LIB "_md"
+#    elif defined(_MT)
+#        define PYBIND11_BUILD_LIB "_mt" // catches /MT or /MTd
 #    else
+#        define PYBIND11_BUILD_LIB ""
+#    endif
+#    if (_MSC_VER) / 100 == 19
+#        define PYBIND11_BUILD_ABI NB_BUILD_LIB "_19"
+#    else
+#        define PYBIND11_BUILD_ABI NB_BUILD_LIB "_unknown"
```

Comment:
I feel strongly that such a fallback will do more harm than good: people will mostly not know this is happening and will get surprised later, possibly after making releases already.

Generating an `#error` here means that we will know immediately that we need to update this code. It'll be under our control then to make this right.

Reply:
As was discussed in PR #4953, MSVC has been ABI-compatible for ages. There is one huge ABI break looming some years in the future (MSVC "vNext") that will mark a transition towards the next island of stability.

Based on the information that is available, this looks to be a painful and disruptive break that will in any case require recompiling all Python packages for a new Windows base platform (i.e., similar to the different "manylinux" flavors). The check here is intended to catch that transition without making pybind11 completely unusable on the vNext version.